src/storage/src/cache/cache_policies/s3_fifo.rs

```diff
@@ -0,0 +1,352 @@
+            *freq = Self::cap_frequency(*freq - 1);
+        }
+    }
+
+    fn inc_small_queue_size(&mut self, size: usize) {
+        self.small_queue_size += size;
+        self.total_size += size;
+    }
+
+    fn dec_small_queue_size(&mut self, size: usize) {
+        self.small_queue_size -= size;
+        self.total_size -= size;
+    }
+
+    fn inc_main_queue_size(&mut self, size: usize) {
+        self.main_queue_size += size;
+        self.total_size += size;
+    }
+
+    fn dec_main_queue_size(&mut self, size: usize) {
+        self.main_queue_size -= size;
+        self.total_size -= size;
+    }
+
+    fn small_queue_fraction(&self) -> f32 {
+        if self.total_size == 0 {
+            0.0
+        } else {
+            self.small_queue_size as f32 / self.total_size as f32
+        }
+    }
+
+    fn check_if_entry_exists_in_small_or_main(&self, entry_id: &EntryID) -> bool {
+        self.frequency.contains_key(entry_id) && !self.ghost_set.contains(entry_id)
+    }
+}
+
+type S3FifoEntrySizeFn = Option<Arc<dyn Fn(&EntryID) -> usize + Send + Sync>>;
+
+/// The policy that implements object size aware S3Fifo algorithm using Deque.
+#[derive(Default)]
+pub struct S3FifoPolicy {
+    state: Mutex<S3FifoInternalState>,
+    size_of: S3FifoEntrySizeFn,
+}
+
+impl fmt::Debug for S3FifoPolicy {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("S3FifoPolicy")
+            .field("state", &self.state)
+            .finish()
+    }
+}
+
+unsafe impl Send for S3FifoPolicy {}
+unsafe impl Sync for S3FifoPolicy {}
+
+impl S3FifoPolicy {
+    /// Create a new [`S3FifoPolicy`].
+    pub fn new(size_of: S3FifoEntrySizeFn) -> Self {
+        Self {
+            state: Mutex::new(S3FifoInternalState::default()),
+            size_of,
+        }
+    }
+
+    fn entry_size(&self, entry_id: &EntryID) -> usize {
+        self.size_of.as_ref().map(|f| f(entry_id)).unwrap_or(1)
+    }
+
+    fn evict_from_small(&self, state: &mut S3FifoInternalState) -> Option<EntryID> {
+        if let Some(element) = state.small.pop_back() {
+            let freq = *state.frequency.get_mut(&element).unwrap_or(&mut 0);
+            let entry_size = self.entry_size(&element);
+            state.dec_small_queue_size(entry_size);
+
+            // If frequency is greater than one, give it another chance, by moving to main queue
+            if freq > 1 {
+                state.main.push_front(element);
+                state.inc_main_queue_size(entry_size);
+                None
+            } else {
+                // Move it to ghost queue
+                state.ghost.push_front(element);
+                state.ghost_set.insert(element);
+                state.frequency.remove(&element);
+                Some(element)
+            }
+        } else {
+            None
+        }
+    }
+
+    fn evict_from_main(&self, state: &mut S3FifoInternalState) -> Option<EntryID> {
+        if let Some(element) = state.small.pop_back() {
```

Comment:
1. Should this be `state.main.pop_back()`?

2. Can we add a test to check that it should pop from `main`?

3. Minor: we can write like `let element = state.main.pop_back()?;` to simplify the if else